### DISCUSSION TITLE
Can I call Write-Message?

### DISCUSSION THREAD

**Discussion by ghost** [dataplat/dbatools#7803] [Q&A] [answered]
Hi, I want to request a feature update for Invoke-DbaDbUpgrade.  I copied the current code and have started to update it to include the new feature.  I can hack out the Write-Message code, but I would like to keep it to see the buried internal exception that it generates to verify it really works.  Is there a way I can either use the Write-Message function or copy it's code to have my own version?  I looked for Write-Message, but can't find it's code.  Perhaps it's compiled in a DLL or something.  

BTW, the refresh view stuff in Invoke-DbaDbUpgrade alerted me to an issue in an upgrade for some views.  I really appreciated the heads up.  I was thinking that I need to test procedures as well.  We used signed procedures sometimes, so added code is required to skip those and issue a warning.

**Comment by MarkIannucci** [comment, marked as answer]
@RandyInMarin , I think you're asking how can I call `Write-Message` locally.  I'm not sure how to do that either.  I was able to find what [appears to be the code for it](https://github.com/sqlcollaborative/dbatools/blob/development/bin/projects/dbatools/dbatools/Commands/WriteMessageCommand.cs) in a binary folder.

**Reply by ghost** [reply]
That's a lot of code to write a message.  I had hoped I could copy and add the new feature without too much hacking.  I will replace write-message and do a simple loop to get the innermost exception.  This might be a feature request to allow code to run outside the module.  Thanks for finding the location.

**Reply by MarkIannucci** [reply]
Hey @RandyInMarin , I don't know if this is much help, but [I asked a similar question (and forgot I did so)](https://github.com/sqlcollaborative/dbatools/discussions/7037).  The answer indicates that the write-message functionality is loosely based on the [Write-PSFMessage](https://psframework.org/documentation/commands/PSFramework/Write-PSFMessage.html) command in the PSFramework module, so perhaps you can use that for your goals?

**Reply by ghost** [reply]
Yes, I think that will do.  Actually, it might replace a bunch of junk I've being doing for logging.
